File: src/Wisp2Mac/Sources/Wisp2Mac/AppLog.swift
import Foundation

enum AppLog {
    static func info(_ area: String, _ message: String) {
        write("INFO", area, message)
    }

    static func error(_ area: String, _ error: Error) {
        write("ERROR", area, error.localizedDescription)
    }

    private static func write(_ level: String, _ area: String, _ message: String) {
        do {
            try AppPaths.ensure()
            let line = "\(Date().formatted(.iso8601)) [\(level)] \(area): \(message)\n"
            if FileManager.default.fileExists(atPath: AppPaths.logURL.path) {
                let handle = try FileHandle(forWritingTo: AppPaths.logURL)
                handle.seekToEndOfFile()
                handle.write(Data(line.utf8))
                handle.closeFile()
            } else {
                try line.write(to: AppPaths.logURL, atomically: true, encoding: .utf8)
            }
        } catch {
            // Logging must never break dictation.
        }
    }
}

File: src/Wisp2Win/Services/AppLog.cs
using System;
using System.IO;

namespace Wisp2Win.Services;

public static class AppLog
{
    private static readonly object SyncRoot = new();

    public static void Info(string area, string message) => Write("INFO", area, message);

    public static void Error(string area, Exception exception) =>
        Write("ERROR", area, $"{exception.GetType().Name}: {exception.Message}\n{exception}");

    public static void Error(string area, string message) => Write("ERROR", area, message);

    private static void Write(string level, string area, string message)
    {
        try
        {
            AppPaths.Ensure();
            var line = $"{DateTimeOffset.Now:O} [{level}] {area}: {message}{Environment.NewLine}";
            lock (SyncRoot)
            {
                File.AppendAllText(AppPaths.LogPath, line);
            }
        }
        catch
        {
            // Logging must never break dictation.
        }
    }
}
